src/Juvix/Core/FromFrontend.hs

```diff
@@ -232,56 +276,125 @@ transformPatVar p = pure p
+  FE.Expression ->
+  Env primTy primVal (HR.Term primTy primVal)
+transformTermHR _ (FE.Constant k) = HR.Prim <$> paramConstant k
+transformTermHR q (FE.Let l) = transformSimpleLet q l
+transformTermHR _ e@(FE.LetType _) = throwFF $ ExprUnimplemented e
+transformTermHR _ e@(FE.Match _) = throwFF $ ExprUnimplemented e
+transformTermHR _ (FE.Name n) = pure $ HR.Elim $ HR.Var n
+transformTermHR q (FE.Lambda l) = transformSimpleLambda q l
+transformTermHR q (FE.Application app) = transformApplication q app
+transformTermHR _ (FE.Primitive (FE.Prim p)) = do
   param <- ask @"param"
   maybe (throwFF $ UnknownPrimitive p) pure $
     primTy param p <|> primVal param p
   where
     primTy param p = HR.PrimTy <$> HM.lookup p (P.builtinTypes param)
     primVal param p = HR.Prim <$> HM.lookup p (P.builtinValues param)
-transformTermHR (FE.List l) =
+transformTermHR _ (FE.List l) =
   throwFF $ ListUnimplemented l
-transformTermHR (FE.Tuple (FE.TupleLit es)) =
-  makeTuple <$> traverse transformTermHR es
-transformTermHR (FE.Block (FE.Bloc e)) =
-  transformTermHR e
-transformTermHR (FE.ExpRecord r) =
+transformTermHR q (FE.Tuple (FE.TupleLit es)) =
+  makeTuple <$> traverse (transformTermHR q) es
+transformTermHR q (FE.Block (FE.Bloc e)) =
+  transformTermHR q e
+transformTermHR _ (FE.ExpRecord r) =
   throwFF $ ExpRecordUnimplemented r
-transformTermHR (FE.ArrowE a) = transformArrow a
-transformTermHR (FE.NamedTypeE (FE.NamedType' _ e)) =
+transformTermHR q (FE.ArrowE a) = transformArrow q a
+transformTermHR q (FE.NamedTypeE (FE.NamedType' _ e)) =
   -- TODO the name will only become relevant (outside of arrows)
   -- when refinements are supported
-  transformTermHR e
-transformTermHR (FE.RefinedE r) =
+  transformTermHR q e
+transformTermHR _ (FE.RefinedE r) =
   throwFF $ RefinementsUnimplemented r
-transformTermHR (FE.UniverseName i) =
+transformTermHR _ (FE.UniverseName i) =
   -- TODO for universe polymorphism
   throwFF $ UniversesUnimplemented i
-transformTermHR (FE.Parened e) = transformTermHR e
-
-transformSimpleLet :: FE.Let -> Env primTy primVal (HR.Term primTy primVal)
-transformSimpleLet e@(FE.LetGroup name (clause :| []) body) = do
+transformTermHR q (FE.Parened e) = transformTermHR q e
+transformTermHR q (FE.DeclarationE (FE.DeclareExpression _ e)) =
+  transformTermHR q e
+
+transformApplication ::
+  NameSymbol.Mod ->
+  FE.Application ->
+  Env primTy primVal (HR.Term primTy primVal)
+transformApplication q (FE.App f xs) =
+  getSpecialE q f >>= flip go (toList xs)
+  where
+    go s xs = case s of
+      Just (ArrowS Nothing) -> do
+        ~[π, a, b] <- nargs 3 xs
+        π <- transformUsage q π
+        go (Just (ArrowS (Just π))) [a, b]
+      Just (ArrowS (Just π)) -> do
+        ~[xa, b] <- nargs 2 xs
+        (x, a) <- namedArg q xa
+        HR.Pi π x a <$> transformTermHR q b
+      Just (PairS Nothing) -> do
+        ~[π, a, b] <- nargs 3 xs
+        π <- transformUsage q π
+        go (Just (PairS (Just π))) [a, b]
```

Comment:
This seems a bit odd, what if we have
```
b : B
f : B -> A
(f b : A)
```

This now seems ambi instead of binding `f b` or `b` to type `A`

Reply:
Further I can follow this up, if A is C -> D by
```
(f b : A) c
```

Reply:
Sorry I don't think I understand the problem. This alternative interpretation is only used after `%Builtin.Pair` (or `Arrow`), in something like `dpair 1 (x: A) (B x)`, and only if the `x` part is a single identifier. In other cases `(x: A)` is just a type annotation (or more precisely an application of `(:)`). Could you clarify?

Reply:
well consider the expression
```haskell

sig f = int -> int -> int
let f = (+)

-- I forget the exact syntax that F* uses for refinements....
let nat = x : int { x >= 0 }

let foo a b = 
  (f a : nat -> nat) a b
```

to me it's not clear that `:` is only used in a signature context so saying that they are Π always seems odd to me?